### ISSUE TITLE
node_agent handshake failure whith mesh expansion

### ISSUE THREAD

**Issue by Docteur-RS** [istio/istio#17886]
This is most probably an error on my end. Please help if you can ;-)  
Thanks !  

**Bug description**

I am following the tutorial for [Istio mesh expansion](https://istio.io/docs/examples/mesh-expansion/single-network/).  
Unfortunatly, when I launch the node_agent I get the following error:  
```
$ sudo node_agent --ca-address istio-citadel:31259 --env onprem
2019-10-15T13:43:25.027036Z     info    parsed scheme: ""
2019-10-15T13:43:25.028755Z     info    scheme "" not registered, fallback to default scheme
2019-10-15T13:43:25.030166Z     info    ccResolverWrapper: sending update to cc: {[{istio-citadel:312590  <nil>}] <nil>}
2019-10-15T13:43:25.031256Z     info    ClientConn switching balancer to "pick_first"
2019-10-15T13:43:25.032732Z     info    Starting Node Agent
2019-10-15T13:43:25.033914Z     info    Node Agent starts successfully.
2019-10-15T13:43:25.032902Z     info    pickfirstBalancer: HandleSubConnStateChange: 0xc000056020, CONNECTING
2019-10-15T13:43:25.268172Z     info    Sending CSR (retrial #0) ...
2019-10-15T13:43:25.397958Z     info    grpc: addrConn.createTransport failed to connect to {istio-citadel:31259 0  <nil>}. Err :connection error: desc = "transport: authentication handshake failed: read tcp10.10.2.99:46821->10.10.1.11:31259: read: connection reset by peer". Reconnecting...
2019-10-15T13:43:25.398088Z     info    pickfirstBalancer: HandleSubConnStateChange: 0xc000056020, TRANSIENT_FAILURE
2019-10-15T13:43:25.398181Z     error   CSR signing failed: rpc error: code = Unavailable desc = all SubConns are in TransientFailure, latest connection error: connection error: desc = "transport: authentication handshake failed: read tcp 10.10.2.99:46821->10.10.1.11:31259: read: connection reset by peer". Will retry in 5s
2019-10-15T13:43:26.398607Z     info    pickfirstBalancer: HandleSubConnStateChange: 0xc000056020, CONNECTING
2019-10-15T13:43:26.846985Z     info    grpc: addrConn.createTransport failed to connect to {istio-citadel:31259 0  <nil>}. Err :connection error: desc = "transport: authentication handshake failed: read tcp10.10.2.99:46822->10.10.1.11:31259: read: connection reset by peer". Reconnecting...
2019-10-15T13:43:26.848962Z     info    pickfirstBalancer: HandleSubConnStateChange: 0xc000056020, TRANSIENT_FAILURE
2019-10-15T13:43:28.582513Z     info    pickfirstBalancer: HandleSubConnStateChange: 0xc000056020, CONNECTING
2019-10-15T13:43:28.778769Z     info    grpc: addrConn.createTransport failed to connect to {istio-citadel:31259 0  <nil>}. Err :connection error: desc = "transport: authentication handshake failed: read tcp10.10.2.99:46823->10.10.1.11:31259: read: connection reset by peer". Reconnecting...
2019-10-15T13:43:28.778907Z     info    pickfirstBalancer: HandleSubConnStateChange: 0xc000056020, TRANSIENT_FAILURE
```
---
* I have copied the certificates from my Kubernetes master to the Debian machine where I installed the node_agent. They are stored in `/etc/certs/` as the docs explains.

* Using node_agent 1.3.2

* Istio has meshExpansion: true

* Content of `cluster.env`
```INI
ISTIO_CP_AUTH=MUTUAL_TLS
ISTIO_SERVICE_CIDR=10.233.0.0/18
```

**Affected product area (please put an X in all that apply)**

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[X] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[X] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

**Expected behavior**
Connection to to the citadel pod through the gateway should work without handshake error

**Steps to reproduce the bug**
Following the tutorial from the documentation step by step

**Version (include the output of `istioctl version --remote` and `kubectl version`)**
Istio : 1.2.5
Kubernetes 1.14.3

**How was Istio installed?**
Installed via Helm

**Environment where bug was observed (cloud vendor, OS, etc)**
Kubernetes on premise (centos7)
Trying to add a debian machine to the mesh service via mesh expansion

**Comment by JimmyCYJ** [comment]
Looks like node agent fails to connect Citadel, @incfly would you mind to take a look since it is mesh expansion related?

**Comment by johscheuer** [comment]
@Docteur-RS could you verify that this DR resolve the issue https://github.com/istio/istio/pull/19498 ? We had the same behaviour until we added the DR after this everything worked fine.

**Comment by johscheuer** [comment]
Okay, so I came a little bit further with Istio 1.4.2 everything works fine but with 1.3.6 I get the same errors like above:

```bash
# Citadel logs
2019-12-12T18:42:44.197398Z	info	grpc: Server.Serve failed to complete security handshake from "10.48.1.7:50218": remote error: tls: unknown certificate
2019-12-12T18:42:45.579131Z	info	grpc: Server.Serve failed to complete security handshake from "10.48.1.7:50226": remote error: tls: unknown certificate
```

I do some further testing to find out where the error happens.

**Comment by johscheuer** [comment]
Sadly this was only premature joy it seems like the operator (`istioctl manifest ..`) doesn't install the default DestinationRule for mTLS which means I was talking to citadel without mTLS. When I add the default DestinationRule I get the same behaviour (until I add the DestinationRule form my PR).

So it seems like Citadel is not able to communicate via mTLS.

**Comment by Docteur-RS** [comment]
@johscheuer Ok... Unfortunate !
Nothing to test then.

**Comment by johscheuer** [comment]
Actually you could test the connection with the `DestinationRule` from here: https://github.com/istio/istio/pull/19553

**Comment by peter-fe** [comment]
Is there any movement on this issue?

It seems that the mesh expansion show case in the documentation is expecting to have global.controlPlaneSecurityEnabled=true set, or at least I assumed that, because of this part in the cluster.env
> ISTIO_CP_AUTH=MUTUAL_TLS

Also it was mentioned [here](https://discuss.istio.io/t/citadel-rejects-requests-from-mesh-expansion-node/2124/4)

> Control Plane mutual TLS should be used since we use gateway with sni routing + mtls for vm to reach out to pilot and citadel.

But to be honest it doesn't seem to make any sense to activate mTLS & controlPlaneSecurity, just to throw in some destination rules to disable it again.

Currently using mesh-expansion + controlPlaneSecurityEnabled is not in a usable condition by following the docs.
Also why I have to provide the certs on the VM when mTLS isn't used.

Is it planed to get mTLS working for mesh-expansion (without disabling it for pilot & citadel) or at least get it into a working state, by adding the merge request done by @johscheuer and skip mTLS for pilot & citadel

**Comment by johscheuer** [comment]
I will try to look into this issue again (was on vacation). I think there is some mismatch in the certificate or the citadel implementation. Also notice that the DR doesn"t deactivate mTLS for the node_agent and citadel (because both are not managed by Istio itself: https://github.com/istio/istio/pull/19553#issuecomment-566175360).

**Comment by Docteur-RS** [comment]
@johscheuer I tried mesh expansion with Istio 1.4 and its still not working.  
I also tried with your PR [#19498](https://github.com/istio/istio/pull/19498) but adding the destination rule doesn't change anyting :/  
I'm still getting certificate errors.  

Using Kubernetes 1.15.3 on premise.

```
 sudo node_agent --ca-address istio-citadel:30716 --env onprem
2020-02-03T13:55:48.812803Z     info    parsed scheme: ""
2020-02-03T13:55:48.812891Z     info    scheme "" not registered, fallback to default scheme
2020-02-03T13:55:48.812910Z     info    ccResolverWrapper: sending update to cc: {[{istio-citadel:30716 0  <nil>}] <nil>}
2020-02-03T13:55:48.812916Z     info    ClientConn switching balancer to "pick_first"
2020-02-03T13:55:48.812946Z     info    Starting Node Agent
2020-02-03T13:55:48.812950Z     info    Node Agent starts successfully.
2020-02-03T13:55:48.816380Z     info    pickfirstBalancer: HandleSubConnStateChange: 0xc000056020, CONNECTING
2020-02-03T13:55:49.053978Z     info    Sending CSR (retrial #0) ...
2020-02-03T13:55:49.178575Z     info    grpc: addrConn.createTransport failed to connect to {istio-citadel:30716 0  <nil>}. Err :connection error: desc = "transport: authentication handshake failed: x509: certificate has expired or is not yet valid". Reconnecting...
2020-02-03T13:55:49.178660Z     info    pickfirstBalancer: HandleSubConnStateChange: 0xc000056020, TRANSIENT_FAILURE
2020-02-03T13:55:49.178722Z     error   CSR signing failed: rpc error: code = Unavailable desc = all SubConns are in TransientFailure, latest connection error: connection error: desc = "transport: authentication handshake failed: x509: certificate has expired or is not yet valid". Will retry in 5s
2020-02-03T13:55:50.179176Z     info    pickfirstBalancer: HandleSubConnStateChange: 0xc000056020, CONNECTING
2020-02-03T13:55:50.253811Z     info    grpc: addrConn.createTransport failed to connect to {istio-citadel:30716 0  <nil>}. Err :connection error: desc = "transport: authentication handshake failed: x509: certificate has expired or is not yet valid". Reconnecting...
2020-02-03T13:55:50.254397Z     info    pickfirstBalancer: HandleSubConnStateChange: 0xc000056020, TRANSIENT_FAILURE
2020-02-03T13:55:51.779190Z     info    pickfirstBalancer: HandleSubConnStateChange: 0xc000056020, CONNECTING
2020-02-03T13:55:52.096071Z     info    grpc: addrConn.createTransport failed to connect to {istio-citadel:30716 0  <nil>}. Err :connection error: desc = "transport: authentication handshake failed: x509: certificate has expired or is not yet valid". Reconnecting...
2020-02-03T13:55:52.096165Z     info    pickfirstBalancer: HandleSubConnStateChange: 0xc000056020, TRANSIENT_FAILURE
```

**Comment by Docteur-RS** [comment]
Steps I did:
```
$ export SERVICE_NAMESPACE="default"
$ kubectl cluster-info dump | grep --color service-cluster-ip-range #To get ip cidr
```
```shell
kubectl -n $SERVICE_NAMESPACE get secret istio.default  \
    -o jsonpath='{.data.root-cert\.pem}' |base64 --decode > root-cert.pem
kubectl -n $SERVICE_NAMESPACE get secret istio.default  \
    -o jsonpath='{.data.key\.pem}' |base64 --decode > key.pem
kubectl -n $SERVICE_NAMESPACE get secret istio.default  \
      -o jsonpath='{.data.cert-chain\.pem}' |base64 --decode > cert-chain.pem

#Installation of certs on the external VM
$ sudo mkdir -p /etc/certs
$ sudo cp {root-cert.pem,cert-chain.pem,key.pem} /etc/certs
```
```
#cluster.env
ISTIO_CP_AUTH=MUTUAL_TLS
ISTIO_SERVICE_CIDR=10.233.0.0/18

$ sudo cp cluster.env /var/lib/istio/envoy
$ sudo chown -R istio-proxy /etc/certs /var/lib/istio/envoy
```

cat `/etc/hosts`  
```
10.10.5.101 istio-citadel istio-pilot istio-pilot.istio-system
```
`sudo node_agent --ca-address istio-citadel:<PORT_CITADEL_VIA_GATEWAY> --env onprem`

---

Istio config:
* meshExpansion.enabled: true
* controlPlaneSecurityEnabled: true

ℹ️  Istio's gateway has been set to NodePort and is accessible from the external VM.

Also, tried with `istio-1.4.3/install/kubernetes\mesh-expansion.yaml`

**Is it possible that the istio gateway absorbs the certificates ?**

**Comment by johscheuer** [comment]
@Docteur-RS these are the steps I did to setup my env: https://github.com/johscheuer/istio-playground/blob/master/Meshexpansion.md sadly I didn't had much time to look into this issue (since I was moved to another project). 

And in my debugging sessions it seems like the gateway somehow (actually not absorbs) validate the cert of Istio Citadel which is missing the SPIFFE URI.

**Comment by JRaver** [comment]
Hi, guys. We have a similar situation with Istio mesh expansion. We have a VM on Vmware on our Datacenters.
We have a Ubuntu 14.04 - 18.04 Os installed on our VMs.
Kubernetes network in our case it's routed, and we can reach pod IP directly.
We successfully added in mesh Ubuntu  14.04 with, but with 18.04 we have a almost similar error like Docteur-RS has.
``` 
2020-02-11T12:53:43.830968Z	info	parsed scheme: ""
2020-02-11T12:53:43.831279Z	info	scheme "" not registered, fallback to default scheme
2020-02-11T12:53:43.831542Z	info	ccResolverWrapper: sending update to cc: {[{istio-citadel:8060 0  <nil>}] <nil>}
2020-02-11T12:53:43.831704Z	info	ClientConn switching balancer to "pick_first"
2020-02-11T12:53:43.831908Z	info	Starting Node Agent
2020-02-11T12:53:43.832010Z	info	Node Agent starts successfully.
2020-02-11T12:53:43.833673Z	info	pickfirstBalancer: HandleSubConnStateChange: 0xc00029c010, CONNECTING
2020-02-11T12:53:44.032995Z	info	Sending CSR (retrial #0) ...
2020-02-11T12:53:45.476998Z	info	grpc: addrConn.createTransport failed to connect to {istio-citadel:8060 0  <nil>}. Err :connection error: desc = "transport: authentication handshake failed: x509: certificate has expired or is not yet valid". Reconnecting...
2020-02-11T12:53:45.477179Z	info	pickfirstBalancer: HandleSubConnStateChange: 0xc00029c010, TRANSIENT_FAILURE
2020-02-11T12:53:45.477241Z	error	CSR signing failed: rpc error: code = Unavailable desc = all SubConns are in TransientFailure, latest connection error: connection error: desc = "transport: authentication handshake failed: x509: certificate has expired or is not yet valid". Will retry in 5s
2020-02-11T12:53:46.477567Z	info	pickfirstBalancer: HandleSubConnStateChange: 0xc00029c010, CONNECTING
2020-02-11T12:53:47.579951Z	info	grpc: addrConn.createTransport failed to connect to {istio-citadel:8060 0  <nil>}. Err :connection error: desc = "transport: authentication handshake failed: x509: certificate has expired or is not yet valid". Reconnecting...
2020-02-11T12:53:47.580003Z	info	pickfirstBalancer: HandleSubConnStateChange: 0xc00029c010, TRANSIENT_FAILURE
2020-02-11T12:53:49.141861Z	info	pickfirstBalancer: HandleSubConnStateChange: 0xc00029c010, CONNECTING 
```


We added the DestinationRule for the citadel with TLS disabled.
But here starts the magic. Sometimes after 2-3 time server can renew cert, sometimes he can't do it.


```
2020-02-11T13:00:28.827711Z	info	parsed scheme: ""
2020-02-11T13:00:28.827746Z	info	scheme "" not registered, fallback to default scheme
2020-02-11T13:00:28.827770Z	info	ccResolverWrapper: sending update to cc: {[{istio-citadel:8060 0  <nil>}] <nil>}
2020-02-11T13:00:28.827779Z	info	ClientConn switching balancer to "pick_first"
2020-02-11T13:00:28.827794Z	info	Starting Node Agent
2020-02-11T13:00:28.827799Z	info	Node Agent starts successfully.
2020-02-11T13:00:28.830984Z	info	pickfirstBalancer: HandleSubConnStateChange: 0xc00002e030, CONNECTING
2020-02-11T13:00:29.089003Z	info	Sending CSR (retrial #0) ...
2020-02-11T13:00:31.766844Z	info	grpc: addrConn.createTransport failed to connect to {istio-citadel:8060 0  <nil>}. Err :connection error: desc = "transport: authentication handshake failed: x509: certificate has expired or is not yet valid". Reconnecting...
2020-02-11T13:00:31.767283Z	info	pickfirstBalancer: HandleSubConnStateChange: 0xc00002e030, TRANSIENT_FAILURE
2020-02-11T13:00:31.767678Z	error	CSR signing failed: rpc error: code = Unavailable desc = all SubConns are in TransientFailure, latest connection error: connection error: desc = "transport: authentication handshake failed: x509: certificate has expired or is not yet valid". Will retry in 5s
2020-02-11T13:00:32.767568Z	info	pickfirstBalancer: HandleSubConnStateChange: 0xc00002e030, CONNECTING
2020-02-11T13:00:33.667898Z	info	grpc: addrConn.createTransport failed to connect to {istio-citadel:8060 0  <nil>}. Err :connection error: desc = "transport: authentication handshake failed: x509: certificate has expired or is not yet valid". Reconnecting...
2020-02-11T13:00:33.668357Z	info	pickfirstBalancer: HandleSubConnStateChange: 0xc00002e030, TRANSIENT_FAILURE
2020-02-11T13:00:35.307261Z	info	pickfirstBalancer: HandleSubConnStateChange: 0xc00002e030, CONNECTING
2020-02-11T13:00:36.366922Z	info	pickfirstBalancer: HandleSubConnStateChange: 0xc00002e030, READY
2020-02-11T13:00:37.306201Z	info	Sending CSR (retrial #1) ...
2020-02-11T13:00:37.310540Z	info	CSR is approved successfully. Will renew cert in 1079h59m59.689788065s
```


On citadel logs

```
2020-02-11T13:00:32.143689Z	info	grpc: Server.Serve failed to complete security handshake from "10.107.4.189:14848": remote error: tls: bad certificate
2020-02-11T13:00:34.044806Z	info	grpc: Server.Serve failed to complete security handshake from "10.107.4.189:14850": remote error: tls: bad certificate
2020-02-11T13:03:06.235994Z	info	transport: loopyWriter.run returning. connection error: desc = "transport is closing"
```

But when we rebooted the servers the cert renews always!!!! For some period of time, yesterday I had rebooted 3 servers with Ubuntu 18.04 and they renew certs, 2 without issues, 1 server with the second try. 
But on the next morning, I restarted the istio-auth-node-agent.service, and all servers can't renew the certs.
We try to look at strace on node_agent and can't find anything. 
We tried with ode_agent 1.3.6,1.3.7, 1.3.8, 1.4.2, 1.4.3 versions, citadel was a 1.3.6,1.3.7, 1.3.8, 1.4.2 versions/

**Comment by Docteur-RS** [comment]
I confirm that the first step woks on Ubuntu 14.04...
Though I cannot establish communication from the VM to the Kubernetes cluster .

Kubernetes => VM ✔️
VM => kubernetes ❌ 

Issues:

# 1)  **istio-sidecar.deb** not installing correctly on ubuntu
When installing using `sudo dpkg -i istio-sidecar.deb` the files are not copied into systemd. I had to extract the .deb and copy the service files to `/etc/systemd/system`.  

# 2) Updating scripts
In **istio-node-agent.sh** I updated the **istio-citadel** port to the nodePort port of my cluster.  
In **istio-start.sh** I updated **pilot** port (to match nodePort again) and **ISTIO_SVC_IP** that was getting its IP from eth0 instead of getting it from eth1.  

# 3) Launching the two scripts using systemctl
node_agent and Istio seems ok but when I do `sudo iptables -L` I get nothing:  
```shell
$ sudo iptables -L
Chain INPUT (policy ACCEPT)
target     prot opt source               destination

Chain FORWARD (policy ACCEPT)
target     prot opt source               destination

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination
```
Its like the iptables script is not doing anything (even when launched manualy).


# 4) Trying to curl the cluster
Curling **myservice** from the VM ends in 503.  
```shell
curl -v myservice.default.svc.cluster.local
* Rebuilt URL to: myservice.default.svc.cluster.local/
* Hostname was NOT found in DNS cache
*   Trying 10.233.35.233...
* Connected to myservice.default.svc.cluster.local (10.233.35.233) port 80 (#0)
> GET / HTTP/1.1
> User-Agent: curl/7.35.0
> Host: myservice.default.svc.cluster.local
> Accept: */*
>
< HTTP/1.1 503 Service Unavailable
< content-length: 91
< content-type: text/plain
< date: Mon, 17 Feb 2020 10:03:50 GMT
* Server envoy is not blacklisted
< server: envoy
<
* Connection #0 to host myservice.default.svc.cluster.local left intact
upstream connect error or disconnect/reset before headers. reset reason: connection failure
```
Strange...  

Also in **istio.err.log** I do have those occasionnal stacktrace:
```
[2020-02-17 10:00:03.483][2652][warning][config] [external/envoy/source/common/config/grpc_mux_subscription_impl.cc:73] gRPC config for type.googleapis.com/envoy.api.v2.Listener rejected: Error adding/updating listener(s) 10.233.27.174_44134: Didn't find a registered implementation for name: 'envoy.listener.http_inspector', 10.233.2.233_6379: Didn't find a registered implementation for name: 'envoy.listener.http_inspector', 10.233.27.187_80: Didn't find a registered implementation for name: 'envoy.listener.http_inspector', 10.233.61.231_50000: Didn't find a registered implementation for name: 'envoy.listener.http_inspector', 10.233.11.160_80: Didn't find a registered implementation for name: 'envoy.listener.http_inspector', 10.233.0.3_9153: Didn't find a registered implementation for name: 'envoy.listener.http_inspector', 0.0.0.0_3000: Didn't find a registered implementation for name: 'envoy.listener.http_inspector', 10.233.27.115_8080: Didn't find a registered implementation for name: 'envoy.listener.http_inspector', 10.233.39.188_443: Didn't find a registered implementation for name: 'envoy.listener.http_inspector', 0.0.0.0_9090: Didn't find a registered implementation for name: 'envoy.listener.http_inspector', 10.233.21.142_443: Didn't find a registered implementation for name: 'envoy.listener.http_inspector', 10.233.0.10_9153: Didn't find a registered implementation for name: 'envoy.listener.http_inspector', 10.233.34.164_80: Didn't find a registered implementation for name: 'envoy.listener.http_inspector', 0.0.0.0_47323: Didn't find a registered implementation for name: 'envoy.listener.http_inspector', 0.0.0.0_9796: Didn't find a registered implementation for name: 'envoy.listener.http_inspector', 10.233.37.168_9000: Didn't find a registered implementation for name: 'envoy.listener.http_inspector', 10.233.49.225_5683: Didn't find a registered implementation for name: 'envoy.listener.http_inspector', 10.233.3.154_80: Didn't find a registered implementation for name: 'envoy.listener.http_inspector', 10.233.57.245_42422: Didn't find a registered implementation for name: 'envoy.listener.http_inspector', 10.233.52.181_3001: Didn't find a registered implementation for name: 'envoy.listener.http_inspector', 10.233.29.203_15020: Didn't find a registered implementation for name: 'envoy.listener.http_inspector'
```

**Comment by Docteur-RS** [comment]
Also tried Istio 1.4.4 and the node_agent doesn't work even on Ubuntu 14...  
So I can't test the interaction between the new .deb (on the VM) and kubernetes because I'm stuck at the CSR signing error.

I tried to merge istio 1.4.3 on the cluster and the files of istio 1.4.4 on the VM ! The results are the following:  
* istio-node-agent-start.sh works (CSR signing etc) ✔️ 
* istio-clean-iptables.sh fails and exits for some unknown reason ❌ 
* istio-start.sh (removed the call to clean-iptables) fails with the following stack trace: ❌ 
In `istio.log`
```
Envoy command: [-c /etc/istio/proxy/envoy-rev0.json --restart-epoch 0 --drain-time-s 45 --parent-shutdown-time-s 60 --service-cluster httpbin --service-node sidecar~10.10.5.251~vagrant.mesh3~mesh3.svc.cluster.local --max-obj-name-len 189 --local-address-ip-version v4 --log-format [Envoy (Epoch 0)] [%Y-%m-%d %T.%e][%t][%l][%n] %v -l debug --component-log-level misc:error]
2020-02-17T14:37:37.591611Z     error   Epoch 0 exited with error: signal: aborted (core dumped)
2020-02-17T14:37:37.591635Z     info    No more active epochs, terminating
```
and in `istio.err.log`
```
[2020-02-17 14:37:49.630][17487][critical][assert] [external/envoy/source/server/hot_restart_impl.cc:45] panic: cannot open shared memory region /envoy_shared_memory_0 check user permissions. Error: File exists
[2020-02-17 14:37:49.630][17487][critical][backtrace] [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:83] Caught Aborted, suspect faulting address 0x6a0000444f
[2020-02-17 14:37:49.630][17487][critical][backtrace] [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:70] Backtrace (use tools/stack_decode.py to get line numbers):
[2020-02-17 14:37:49.630][17487][critical][backtrace] [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:71] Envoy version: 4cd27af28f54d6da1a487aa756256eb5393fff7d/1.12.0/Clean/RELEASE/BoringSSL
[2020-02-17 14:37:49.631][17487][critical][backtrace] [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #0: __restore_rt [0x7fb5fef23330]
[2020-02-17 14:37:49.633][17487][critical][backtrace] [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #1: Envoy::Server::HotRestartImpl::HotRestartImpl() [0x559f8e5b5778]
[2020-02-17 14:37:49.635][17487][critical][backtrace] [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #2: Envoy::MainCommonBase::MainCommonBase() [0x559f8d7595ae]
[2020-02-17 14:37:49.636][17487][critical][backtrace] [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #3: Envoy::MainCommon::MainCommon() [0x559f8d759f3a]
[2020-02-17 14:37:49.638][17487][critical][backtrace] [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #4: main [0x559f8d758c83]
[2020-02-17 14:37:49.638][17487][critical][backtrace] [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #5: __libc_start_main [0x7fb5feb6bf45]
```

**Comment by fscz** [comment]
I ran into this issue a couple of days ago. Initially, I successfully connected a Ubuntu 18.04 VM to Istio 1.4.6 (mesh and sidecar on VM), however the next day, I noticed the error:
"transport: authentication handshake failed: x509: certificate has expired or is not yet valid"

As it turned out, the rotating certificates that Istio generates have somewhat tight validity constraints. They are essentially valid from NOW until some time in the future. In my particular case, the problem was that over-night, the time of my Ubuntu VM had drifted one second. The connection could not be established because the required certificate was only valid, starting from one second into the future. 

I was able to resolve the issue by installing chronyd on my Ubuntu VM and have it sync the system time.
